test/e2e/tests/setup_e2e_test.go

```diff
@@ -68,27 +71,41 @@ var _ = BeforeSuite(func() {
 	Logf("Running in project %v with service account %v\n\n", *project, *serviceAccount)
 
 	for _, zone := range zones {
-		nodeID := fmt.Sprintf("gce-pd-csi-e2e-%s", zone)
-
-		i, err := remote.SetupInstance(*project, zone, nodeID, *serviceAccount, computeService)
-		Expect(err).To(BeNil())
-
-		testInstances = append(testInstances, i)
+		go func(curZone string) {
+			defer GinkgoRecover()
+			nodeID := fmt.Sprintf("gce-pd-csi-e2e-%s", curZone)
+			Logf("Setting up node %s\n", nodeID)
+
+			i, err := remote.SetupInstance(*project, curZone, nodeID, *serviceAccount, computeService)
+			Expect(err).To(BeNil())
+
+			// Create new driver and client
+			testContext, err := testutils.GCEClientAndDriverSetup(i)
+			Expect(err).To(BeNil(), "Set up new Driver and Client failed with error")
+			tcc <- testContext
+		}(zone)
 	}
 
+	for i := 0; i < len(zones); i++ {
+		tc := <-tcc
+		Logf("Test Context for node %s set up\n", tc.Instance.GetName())
+		testContexts = append(testContexts, tc)
+	}
 })
 
 var _ = AfterSuite(func() {
-	/*
-		err := node.client.CloseConn()
-		if err != nil {
-			Logf("Failed to close the client")
-		} else {
-			Logf("Closed the client")
-	*/
-	for _, i := range testInstances {
+
+	for _, tc := range testContexts {
+		err := remote.TeardownDriverAndClient(tc)
+		Expect(err).To(BeNil(), "Teardown Driver and Client failed with error")
 		if *deleteInstances {
-			i.DeleteInstance()
+			tc.Instance.DeleteInstance()
 		}
 	}
 })
+
+func getRandomTestContext() *remote.TestContext {
+	Expect(testContexts).ToNot(BeEmpty())
+	rn := rand.Intn(len(testContexts))
+	return testContexts[rn]
```

Comment:
Is there any issue if multiple tests run in parallel and choose the same context?

Reply:
I don't believe so, since the driver can handle multiple incoming requests and all volume's should be unique due to UUID's. If there is a parallelism issue here it would be with the driver, not the tests and that is something we'd want to catch if it existed.